gcloud/storage/batch.py

```
@@ -0,0 +1,184 @@
+
+    :type connection: :class:`gcloud.storage.connection.Connection`
+    :param connection: the connection for which the batch proxies.
+    """
+    _MAX_BATCH_SIZE = 1000
+
+    def __init__(self, connection):
+        super(Batch, self).__init__(project=connection.project)
+        self._connection = connection
+        self._requests = []
+        self._responses = []
+
+    def _do_request(self, method, url, headers, data):
+        """Override Connection:  defer actual HTTP request.
+
+        Only allow up to ``_MAX_BATCH_SIZE`` requests to be deferred.
+
+        :type method: string
+        :param method: The HTTP method to use in the request.
+
+        :type url: string
+        :param url: The URL to send the request to.
+
+        :type headers: dict
+        :param headers: A dictionary of HTTP headers to send with the request.
+
+        :type data: string
+        :param data: The data to send as the body of the request.
+
+        :rtype: tuple of ``response`` (a dictionary of sorts)
+                and ``content`` (a string).
+        :returns: The HTTP response object and the content of the response.
+        """
+        if method == 'GET':
+            _req = self._connection.http.request
+            return _req(method=method, uri=url, headers=headers, body=data)
+
+        if len(self._requests) >= self._MAX_BATCH_SIZE:
+            raise ValueError("Too many deferred requests (max %d)" %
+                             self._MAX_BATCH_SIZE)
+        self._requests.append((method, url, headers, data))
+        return NoContent(), ''
+
+    def finish(self):
+        """Submit a single `multipart/mixed` request w/ deferred requests.
+
+        :rtype: list of tuples
+        :returns: one ``(status, reason, payload)`` tuple per deferred request.
+        :raises: ValueError if no requests have been deferred.
+        """
+        if len(self._requests) == 0:
+            raise ValueError("No deferred requests")
```

I have it raise because it is a programming error:  a silent failure won't help the app developer find the bug.  For instance, it was helpful to me when I was debugging using batches in the regression tests.